sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -1429,4 +1435,97 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
       case _ =>
     }
   }
+
+  def checkDefaultValuesInPlan(plan: LogicalPlan, isForV1: Boolean = false): Unit = {
+    // Do not check anything if the children are not resolved yet.
+    if (!plan.childrenResolved) return
+    plan match {
+      case AlterColumn(t: ResolvedTable, col: ResolvedFieldName, _, _, _, _,
+          Some(default: DefaultValueExpression)) =>
+        checkTableProvider(t.catalog, t.name, getTableProviderFromProp(t.table.properties()))
+        checkDefaultValue(default, t.name, col.name, col.field.dataType, isForV1)
+
+      case cmd: V2CreateTablePlan if cmd.columns.exists(_.defaultValue.isDefined) =>
+        val ident = cmd.name.asInstanceOf[ResolvedIdentifier]
+        checkTableProvider(ident.catalog, ident.name, cmd.tableSpec.provider)
+        cmd.columns.filter(_.defaultValue.isDefined).foreach { col =>
+          val Column(name, dataType, _, _, Some(default), _) = col
+          // CREATE/REPLACE TABLE only has top-level columns
+          val colName = Seq(name)
+          checkDefaultValue(default, ident.name, colName, dataType, isForV1)
+        }
+
+      case cmd: AlterTableCommand =>
+        val table = cmd.table.asInstanceOf[ResolvedTable]
+        cmd.transformExpressionsDown {
+          case q @ QualifiedColType(path, Column(name, dataType, _, _, Some(default), _), _)
+              if path.resolved =>
+            checkTableProvider(
+              table.catalog, table.name, getTableProviderFromProp(table.table.properties()))
+            checkDefaultValue(
+              default,
+              table.name,
+              path.name :+ name,
+              dataType,
+              isForV1)
+            q
+        }
+
+      case _ =>
+    }
+  }
+
+  private def getTableProviderFromProp(props: JMap[String, String]): Option[String] = {
+    Option(props.get(TableCatalog.PROP_PROVIDER))
+  }
+
+  private def checkTableProvider(
+      catalog: CatalogPlugin,
+      tableName: String,
+      provider: Option[String]): Unit = {
+    // We only need to check table provider for the session catalog. Other custom v2 catalogs
+    // can check table providers in their implementations of createTable, alterTable, etc.
+    if (!CatalogV2Util.isSessionCatalog(catalog)) return
```

Comment:
same as above, we should probably reverse the logic and put the below in the `if` block body instead.